clients/drcachesim/tools/opcode_mix.cpp

```diff
@@ -338,5 +341,116 @@ opcode_mix_t::print_results()
     return true;
 }
 
+opcode_mix_t::interval_state_snapshot_t *
+opcode_mix_t::generate_interval_snapshot(uint64_t interval_id)
+{
+    return generate_shard_interval_snapshot(&serial_shard_, interval_id);
+}
+
+opcode_mix_t::interval_state_snapshot_t *
+opcode_mix_t::generate_shard_interval_snapshot(void *shard_data, uint64_t interval_id)
+{
+    assert(shard_data != nullptr);
+    auto &shard = *reinterpret_cast<shard_data_t *>(shard_data);
+    auto *snap = new snapshot_t;
+    snap->instr_count_ = shard.instr_count;
+    snap->opcode_counts_ = shard.opcode_counts;
+    snap->category_counts_ = shard.category_counts;
+    return snap;
+}
+
+bool
+opcode_mix_t::finalize_interval_snapshots(
+    std::vector<interval_state_snapshot_t *> &interval_snapshots)
+{
+    // Loop through snapshots in reverse order, subtracting the *earlier*
+    // snapshot's cumulative values from this snapshot's cumulative values, to get
+    // deltas.  The first snapshot needs no updates, obviously.
+    for (int i = static_cast<int>(interval_snapshots.size()) - 1; i > 0; --i) {
+        auto &this_snap = *reinterpret_cast<snapshot_t *>(interval_snapshots[i]);
+        const auto &prior_snap =
+            *reinterpret_cast<const snapshot_t *>(interval_snapshots[i - 1]);
+        this_snap.instr_count_ -= prior_snap.instr_count_;
+        for (auto &opc_count : this_snap.opcode_counts_) {
+            auto prior_opc_it = prior_snap.opcode_counts_.find(opc_count.first);
+            if (prior_opc_it != prior_snap.opcode_counts_.end()) {
+                opc_count.second -= prior_opc_it->second;
+            }
+        }
+        for (auto &cat_count : this_snap.category_counts_) {
+            auto prior_cat_it = prior_snap.category_counts_.find(cat_count.first);
+            if (prior_cat_it != prior_snap.category_counts_.end()) {
+                cat_count.second -= prior_cat_it->second;
+            }
+        }
+    }
+    return true;
+}
+
+opcode_mix_t::interval_state_snapshot_t *
+opcode_mix_t::combine_interval_snapshots(
+    const std::vector<const interval_state_snapshot_t *> latest_shard_snapshots,
+    uint64_t interval_end_timestamp)
+{
+    snapshot_t *super_snap = new snapshot_t;
+    for (const interval_state_snapshot_t *base_snap : latest_shard_snapshots) {
+        const auto *snap = reinterpret_cast<const snapshot_t *>(base_snap);
```

Comment:
Looks like your interval snapshots contain deltas. Note that comment here about combining them correctly: https://github.com/DynamoRIO/dynamorio/blob/4c8e947dcf33fab66c25c737297a9230a3bc7a9f/clients/drcachesim/analysis_tool.h#L375.

Reply:
Huh, I had to read that comment a bunch of times to understand why you're pointing this out.  The fact that latest_shard_snapshots may contains snapshots from prior intervals is a fairly subtle property I hadn't appreciated before.  I don't think it's ever directly stated.  But good catch, I'll have to fix my code to filter out the irrelevant snapshots.